[specific ci=Group1-Docker-Commands] Optimize Docker Command tests to one VCH

sflxn opened the pull request:
Migrate the tests to using a single VCH if running as a group.
There are still some test files that install their a new VCH.
Some of those tests should be re-examined to determine if they
really need a new VCH.

A bootstrap and cleanup robot file is required to initiate and
clean up a group of tests.  The infrastructure does not prevent
single test file from running normally.

To initiate "one VCH" mode, REUSE-VCH env variable must be
set to the VCH name in the bootstrap file.  Also, the VCH name
must be added to the removal exception list.

1. Added exception list to prevent VCH from getting "cleaned up"
2. Modify cleanup code to ignore assets linked to the exception list
3. Clean up containers with docker before cleaning up VIC appliance
4. Fix tests with dependencies on it running on its own VCH
5. Added a few docker utility keywords
6. Add DEBUG_VCH secret to allow using an existing VCH created outside of CI

Resolves #6301

anchal-agrawal commented on tests/resources/Docker-Util.robot:
```diff
@@ -201,4 +201,86 @@ Verify Volume Inspect Info
+    ${rc}  ${out}=  Run And Return Rc And Output  docker %{VCH-PARAMS} ps -a
+    Run Keyword If  ${rc} != 0  Log To Console  Remaining containers - ${out}
+
+Does Images Exist
```
s/Does/Do

anchal-agrawal commented on tests/resources/Docker-Util.robot:
```diff
@@ -201,4 +201,86 @@ Verify Volume Inspect Info
+    Return From Keyword If  ${len} == 0  ${false}
+    [Return]  ${true}
+
+Does Containers Exist
```
s/Does/Do

anchal-agrawal commented on tests/resources/Docker-Util.robot:
```diff
@@ -201,4 +201,86 @@ Verify Volume Inspect Info
+    Return From Keyword If  ${len} == 0  ${false}
+    [Return]  ${true}
+
+Does Volumes Exist
```
s/Does/Do

anchal-agrawal commented on tests/resources/Docker-Util.robot:
```diff
@@ -201,4 +201,86 @@ Verify Volume Inspect Info
+    Return From Keyword If  ${len} == 0  ${false}
+    [Return]  ${true}
+
+Does Networks Exist
```
s/Does/Do

anchal-agrawal commented on tests/resources/Docker-Util.robot:
```diff
@@ -201,4 +201,86 @@ Verify Volume Inspect Info
+    Return From Keyword If  ${exist} == ${false}  0
+
+    Run Keyword If  ${exist}  Kill All Containers
+    Run Keyword If  ${exist}  Run  docker %{VCH-PARAMS} rm $(docker %{VCH-PARAMS} ps -a -q)
```
Should we simply `rm -f` containers here, just to be sure? That way we can skip the `kill` step too.

anchal-agrawal commented on tests/resources/VCH-Util.robot:
```diff
@@ -160,8 +160,66 @@ Get Docker Params
+    Set Environment Variable  vm-exceptions  ${list-string}
+    Log To Console  Saved '${list-string}' to removal exceptions
+
+Check If VCH is In Exception
```
Minor: inconsistent casing - `Check If VCH is In Exception`. We could say `Is In` or `Check if VCH is in Exception`.

anchal-agrawal commented on tests/resources/VCH-Util.robot:
```diff
@@ -160,8 +160,66 @@ Get Docker Params
+
 Install VIC Appliance To Test Server
     [Arguments]  ${vic-machine}=bin/vic-machine-linux  ${appliance-iso}=bin/appliance.iso  ${bootstrap-iso}=bin/bootstrap.iso  ${certs}=${true}  ${vol}=default  ${cleanup}=${true}  ${debug}=1  ${additional-args}=${EMPTY}
+    # Don't install if reuse is true and a vch exist
```
Is this comment not needed?

anchal-agrawal commented on tests/test-cases/Group1-Docker-Commands/1-00-Bootstrap.robot:
```diff
@@ -0,0 +1,26 @@
+Setup Docker Commands Tests
+    Set Environment Variable  REUSE-VCH  %{VCH-NAME}
+    Add VCH to Removal Exception List  vch=%{VCH-NAME}
+    Add VCH to Removal Exception List  vch=test1
```
What is `test1` here?

anchal-agrawal commented on tests/test-cases/Group1-Docker-Commands/1-00-Bootstrap.robot:
```diff
@@ -0,0 +1,26 @@
+    Set Environment Variable  REUSE-VCH  %{VCH-NAME}
+    Add VCH to Removal Exception List  vch=%{VCH-NAME}
+    Add VCH to Removal Exception List  vch=test1
+    Remove VCH From Removal Exception List  vch=test1
```
Why are we adding and removing `test1` here?

anchal-agrawal commented on tests/test-cases/Group1-Docker-Commands/1-03-Docker-Images.robot:
```diff
@@ -30,19 +30,19 @@ Simple images
     Should Be Equal As Integers  ${rc}  0
     Should Not Contain  ${output}  Error
-    ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} images
+    ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} images alpine
```
Is this necessary because the shared VCH will have more images? Would this shared VCH have `alpine` more than 3 times?

sflxn replied:
yes, because we have one vch and there are other images.

anchal-agrawal commented on tests/test-cases/Group1-Docker-Commands/1-10-Docker-PS.robot:
```diff
@@ -158,13 +161,17 @@ Docker ps Remove container OOB
     Run Keyword If  '%{DATASTORE_TYPE}' == 'VSAN'  Should Be Equal As Integers  ${rc}  0
     ${rc}  ${output}=  Run Keyword If  '%{DATASTORE_TYPE}' == 'Non_VSAN'  Run And Return Rc And Output  govc datastore.ls | grep ${container} | xargs -n1 govc datastore.rm
     Run Keyword If  '%{DATASTORE_TYPE}' == 'Non_VSAN'  Should Be Equal As Integers  ${rc}  0
+    # ${rc}  ${out}=  Run And Return Rc And Output  docker %{VCH-PARAMS} ps
```
Should this comment and the ones below be removed?

anchal-agrawal commented on tests/test-cases/Group1-Docker-Commands/1-10-Docker-PS.robot:
```diff
@@ -108,6 +108,9 @@ Docker ps powerOn container OOB
 
     Wait Until Keyword Succeeds  10x  6s  Assert Number Of Containers  ${len+1}
 
+    ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} stop jojo
```
Why is this step needed?

sflxn replied:
There is a bug with docker kill and OOB stopped containers.

anchal-agrawal commented on tests/test-cases/Group1-Docker-Commands/1-20-Docker-Volume-Inspect.robot:
```diff
@@ -29,7 +29,7 @@ Simple docker volume inspect
-    Should Be Equal As Integers  ${rc}  1
-    Should Contain  ${output}  Error: No such volume: fakeVolume
\ No newline at end of file
+# Docker volume inspect invalid object
```
We should uncomment this test.

anchal-agrawal commented on tests/resources/Docker-Util.robot:
```diff
@@ -201,4 +201,86 @@ Verify Volume Inspect Info
+    Return From Keyword If  ${exist} == ${false}  0
+    [Return]  1
+
+Remove All Networks
```
Suggest renaming this to `Remove All Container Networks` for clarity.

anchal-agrawal commented on tests/resources/Docker-Util.robot:
```diff
@@ -201,4 +201,86 @@ Verify Volume Inspect Info
+
+Remove All Networks
+    ${exist}=  Does Networks Exist
+    Run Keyword If  ${exist}  Log To Console  Removing all networks from %{VCH-NAME}
```
s/Removing all networks from/Removing all container networks from

sgairo commented on tests/README.md:
```diff
@@ -59,3 +67,16 @@ Use ./local-integration-test.sh
+    tests to run in either shared VCH or standalone VCH mode.
+    - Add a cleanup.robot file that handles cleaning up the group-wide VCH.  It needs to remove the group-wide VCH-NAME from the cleanup exception list.
+2. Write all tests within robot file with the assumption that the VCH is in shared mode.  Don't assume there are no previously created containers and images.  If a robot file needs this precondition, make sure the suite setup cleans out the VCH before running any test.
+3. If there is an existing VCH available, it is possible to bypass the VCH installation/delection by adding a DEBUG_VCH into the list of test secrets.
```
"delection" -> "deletion"

anchal-agrawal commented on tests/test-cases/Group1-Docker-Commands/1-10-Docker-PS.robot:
```diff
@@ -108,6 +108,9 @@ Docker ps powerOn container OOB
 
     Wait Until Keyword Succeeds  10x  6s  Assert Number Of Containers  ${len+1}
 
+    ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} stop jojo
```
I'm not sure why this is needed still... how does having a shared VCH require this? If there's a bug here, does it not exist in our current integration test suite?

sflxn replied:
There is a bug that hasn't been caught before.  To get around this bug, I must stop the container here.  The PS test doesn't clean up their tests, and I have a clean up routine that tries to stop and delete all containers at the end of each suite, but it fails because we stopped the container out of band here.

mhagen-vmware commented on tests/test-cases/Group1-Docker-Commands/1-03-Docker-Images.robot:
```diff
@@ -30,19 +30,19 @@ Simple images
     Should Be Equal As Integers  ${rc}  0
     Should Not Contain  ${output}  Error
-    ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} images
+    ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} images alpine
```
This is really fundamentally changing the nature of this test

sflxn replied:
No, it doesn't.  This test pulls down 3 versions of alpines and checks if docker images shows 3 alpine images.  That's what the above does.

mhagen-vmware commented on tests/test-cases/Group1-Docker-Commands/1-03-Docker-Images.robot:
```diff
@@ -30,19 +30,19 @@ Simple images
 
 All images
-    ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} images -a
+    ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} images -a alpine
```
This is fundamentally changing the nature of this test

sflxn replied:
Again, it doesn't.  This test assumes an independent VCH.  It pulls down 3 versions of Alpines and want to see if docker images -a returns 3 images.  Adding a name to docker images doesn't simply filters images by that name.

mlh78750 commented:
This changes the coverage and tests these operations in a new way (with just one VCH). This removes coverage for the same tests given a new VCH each time (initial operation). Can we look at running these both ways so that we get coverage of each type of use?

sflxn commented:
@mlh78750 @mhagen-vmware I've added another variable MULT_VCH that enables the older mode of operations where each suite installs its own VCH and cleans it up.  If it's not set or set to anything but 1, the tests will assume a shared VCH mode.

TARGET_VCH was also added.  That ignores MULTI_VCH.  If that is set, the test will use that target vch as the shared vch.

DEBUG_VCH now enables logging.  It dumps docker images and running containers on a VCH before the run of a suite.
